massa-bootstrap/src/server.rs

```diff
@@ -14,292 +44,506 @@ use massa_signature::KeyPair;
+    };
+
+    let listener = establisher
+        .get_listener(listen_addr)
+        .await
+        .map_err(BootstrapError::IoError)?;
+
+    // This is the primary interface between the async-listener, and the "sync" worker
+    let (listener_tx, listener_rx) = crossbeam::channel::bounded::<BsConn>(max_bootstraps * 2);
+
+    let white_black_list = SharedWhiteBlackList::new(
+        config.bootstrap_whitelist_path.clone(),
+        config.bootstrap_blacklist_path.clone(),
+    )?;
+
+    let updater_lists = white_black_list.clone();
+    let update_handle = thread::Builder::new()
+        .name("wb_list_updater".to_string())
+        .spawn(move || {
+            let res = BootstrapServer::run_updater(
+                updater_lists,
+                config.cache_duration.into(),
+                update_stopper_rx,
+            );
+            match res {
+                Ok(_) => info!("ip white/blacklist updater exited cleanly"),
+                Err(ref er) => error!("updater exited with error: {}", er),
+            };
+            res
+        })
+        // the non-builder spawn doesn't return a Result, and documentation states that
+        // it's an error at the OS level.
+        .unwrap();
+    let listen_rt_handle = bs_server_runtime.handle().clone();
+    let listen_handle = thread::Builder::new()
+        .name("bs_listener".to_string())
+        .spawn(move || {
+            let res =
+                listen_rt_handle.block_on(BootstrapServer::run_listener(listener, listener_tx));
+            res
+        })
+        // the non-builder spawn doesn't return a Result, and documentation states that
+        // it's an error at the OS level.
+        .unwrap();
+
+    let main_handle = thread::Builder::new()
+        .name("bs-main-loop".to_string())
+        .spawn(move || {
             BootstrapServer {
                 consensus_controller,
                 network_command_sender,
                 final_state,
-                establisher,
-                manager_rx,
-                bind,
+                listener_rx,
+                listen_stopper_rx,
+                white_black_list,
                 keypair,
                 version,
-                ip_hist_map: HashMap::with_capacity(bootstrap_config.ip_list_max_size),
-                bootstrap_config,
+                ip_hist_map: HashMap::with_capacity(config.ip_list_max_size),
+                bootstrap_config: config,
+                bs_server_runtime,
             }
-            .run()
-            .await
-        });
-        Ok(Some(BootstrapManager {
-            join_handle,
-            manager_tx,
-        }))
-    } else {
-        Ok(None)
-    }
+            .run_loop(max_bootstraps)
+        })
+        // the non-builder spawn doesn't return a Result, and documentation states that
+        // it's an error at the OS level.
+        .unwrap();
```

Comment:
we use `expect("explanation on where this happened")` and the error is automatically appended. That way we know where the panic happened even when there is no traceback

Reply:
#3693

Reply:
ok for the followup